{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter FASTA or FASTQ file to read. If this request was a mistake, press enter with no other input. Otherwise, enter a file containing DNA sequences: \n",
      "human_g1k_v37_truncated.fasta\n",
      "Write the path of the file you want to output the information to! If this was a mistake, presss enter. Otherwise, provide the file path: \n",
      "'tre\n",
      "Outputting information to tre...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class NotFastaOrFastqFileError(Exception): \n",
    "    \"\"\"This class is an exception will be generated if the file type is incorrect.\n",
    "       The instantiation is overriden and a value (the file extension) is passed to an instance variable.\n",
    "       This is instance variable 'value' is returned when a string representation is requested. \n",
    "    \"\"\"\n",
    "    def __init__(self, value):\n",
    "        self.value = value\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return repr(self.value)\n",
    "\n",
    "\"\"\"This function gets the file extension from the filePath. If no file \n",
    "    extension is found, returns the message \"Not a file\". This is what will\n",
    "    be passed to the exception handling if it is not a FASTA or FASTQ file.\n",
    "    \n",
    "\"\"\"\n",
    "def getFileExtension(filePath):\n",
    "    if '.' in filePath:\n",
    "        \"\"\"Reverses the filePath and then finds the first (i.e., the last)'.' in the reversed\n",
    "            filePath. \"\"\"\n",
    "        extensionPeriodLocation = filePath[::-1].find('.')\n",
    "\n",
    "        \"\"\"The variable fileExtension is only true for the reverse string. This code makes \n",
    "            sure it applies to the standard orientation. Because in a slice the execution\n",
    "            never reaches the item in the middle position, we must add 1 to it so we get \n",
    "            the period. Since we are moving backwards, we are converting it to a negative\n",
    "            number. We then use a second slice operation to get the result facing foward.\n",
    "        \"\"\"\n",
    "        fileExtension = filePath[-1:-(extensionPeriodLocation +1):-1][::-1]\n",
    "        \"\"\"In case there is a period in one of the sections between the \n",
    "            slashes, but the portion after the last slash contains no periods.\n",
    "            In this cae, it is probably a directory if it is a valid path at\n",
    "            all. \"\"\"\n",
    "        if fileExtension != '':\n",
    "            \"\"\"The lowercase function is used in case the Biopython parse \n",
    "            function requires lowercase arguments for the file type. Not sure\n",
    "            how much this matters but good just to be safe. \n",
    "            \"\"\"\n",
    "            return fileExtension.lower()\n",
    "        \n",
    "        else: return \"Not a file\"\n",
    "    # Has no file extension, so cannot  be a file\n",
    "    else:\n",
    "        return \"Not a file\"\n",
    "\n",
    "def formatAminoAcids(count_dictionary):\n",
    "    ### Amino acid names  gottens from http://www.chem.qmul.ac.uk/iupac/AminoAcid/AA1n2.html \n",
    "    # key of new dictionary is amino acid name, value is amino acid count\n",
    "    formatted_dictionary = {'Alanine': count_dictionary['A']}\n",
    "    formatted_dictionary['Arginine'] = count_dictionary['R']\n",
    "    formatted_dictionary['Asparagnine'] = count_dictionary['N']\n",
    "    formatted_dictionary['Aspartic acid'] = count_dictionary['D']\n",
    "    formatted_dictionary['Cysteine'] = count_dictionary['C']\n",
    "    ### 5 done\n",
    "    formatted_dictionary['Glutamine'] = count_dictionary['Q']\n",
    "    formatted_dictionary['Glutamic acid'] = count_dictionary['E']\n",
    "    formatted_dictionary['Glycine'] = count_dictionary['G']\n",
    "    formatted_dictionary['Histidine'] = count_dictionary['H']\n",
    "    formatted_dictionary['Isoleucine'] = count_dictionary['I']\n",
    "    # 10 done\n",
    "    formatted_dictionary['Leucine'] = count_dictionary['L']\n",
    "    formatted_dictionary['Lysine'] = count_dictionary['K']\n",
    "    formatted_dictionary['Methionine'] = count_dictionary['M']\n",
    "    formatted_dictionary['Phenylalanine'] = count_dictionary['F']\n",
    "    formatted_dictionary['Proline'] = count_dictionary['P']\n",
    "    ### 15 done'\n",
    "    formatted_dictionary['Serine'] = count_dictionary['S']\n",
    "    formatted_dictionary['Threonine'] = count_dictionary['T']\n",
    "    formatted_dictionary['Tryptophan'] = count_dictionary['W']\n",
    "    formatted_dictionary['Tyrosine'] = count_dictionary['Y']\n",
    "    formatted_dictionary['Valine'] = count_dictionary['V']\n",
    "    return formatted_dictionary\n",
    "\n",
    "\"\"\" This function actually processes data from the FASTA or FASTQ file. This is pretty much just a loop\n",
    "    that iterates through all the records of the SeqIO object. It converts the input into the necessary\n",
    "    forms required to count the amino acids, and then writes the relevant information to the file specified\n",
    "    by \"output.\" \"iterator\" should be the iterator returned by parsing a SEQIO object.\n",
    "    Using the IUPACAmbiguousDNA Alphabet... our example has N in it, \n",
    "    and it is good because it allows for flexibility in future inputs, as well. This should include\n",
    "    all DNA files we input. It also includes GATC, which are  the only letters  in \n",
    "    IUPACUnambiguousDNA. IUPACAmbiguousDNA allows for a greater subset. \"\"\"\n",
    "def processFile(iterator,output):\n",
    "    ### This loop prints the protein count of all the record###\n",
    "    for record in iterator:\n",
    "        ### gets the sequence from the record\n",
    "        thisSeq = record.seq\n",
    "        '''This prevents a Biopython warning from showing up if there are incomplete codons.\n",
    "            Appends an N (wildcard) to the sequence until it is divisible by 3. This is essentially what is \n",
    "            recommended in the Biopython warning message.'''\n",
    "        if (len(thisSeq) % 3 != 0):\n",
    "            leftoverNucCount = len(thisSeq) % 3 \n",
    "            for i in range(3 - leftoverNucCount):\n",
    "                thisSeq = Seq((str(thisSeq)+ \"N\"), thisSeq.alphabet)\n",
    "        ## translates the record as a new amino acid/peptide sequence\n",
    "        translated_sequence = thisSeq.translate()\n",
    "        ### This changes the sequence to a ProteinAnalysis object, which lets us call the needed methods on it. ###\n",
    "        analyzed_sequence = ProteinAnalysis(str(translated_sequence))\n",
    "        output.write(\"Name: {0}\\nDescription: {1}\\nAnnotations: {2}\".format(record.name, record.description,\n",
    "                                                                            record.annotations))\n",
    "        ### gets the amino acid count\n",
    "        aminoAcidCountDictionary = formatAminoAcids(analyzed_sequence.count_amino_acids())\n",
    "        ### prints the amino acid count!\n",
    "        output.write(\"\\n\\nThis is the amino acid count of record {0}:\".format(record.id) +\"\\n\\n\")\n",
    "        ### splits the output so that each Amino Acid gets it's own line\n",
    "        for aminoAcid, count in aminoAcidCountDictionary.items():\n",
    "            output.write(aminoAcid + \": \" + str(count)+\"\\n\")\n",
    "        ###print(aminoAcidCountDictionary)\n",
    "        ### turns the sequence into  RNA\n",
    "        thisSeqRNA= thisSeq.transcribe()\n",
    "        analyzed_RNAsequence = ProteinAnalysis(str(thisSeqRNA.translate()))\n",
    "\n",
    "        \"\"\" Since the RNA is the same as the DNA with the exception of one nucleotide, getting an amino acid\n",
    "        count from the RNA should be the same as the amino acid count from the DNA.\n",
    "        \"\"\"\n",
    "        output.write(\"\\n\\nThis is the amino acid count of the protein sequnence derived from the RNA resulting from \"\n",
    "        \"the DNA sequence in the file. It should be the same as the previous amino acid count: \\n\\n\")\n",
    "        rnaAcidCountDictionary = formatAminoAcids(analyzed_RNAsequence.count_amino_acids())\n",
    "\n",
    "        for aminoAcid, count in rnaAcidCountDictionary.items():\n",
    "\n",
    "            output.write(aminoAcid+ \": \"+ str(count)+\"\\n\")\n",
    "        output.write('\\n**************************************\\n\\n')\n",
    "\n",
    "\n",
    "while True:\n",
    "    try:\n",
    "    # this gets input from the Python interpreter... We can then pass this to SeqIo to read it into Biopython\n",
    "        fileInputPath = input(\"Enter FASTA or FASTQ file to read. If this request was a mistake, \" \n",
    "        \"press enter with no other input. Otherwise, enter a file containing DNA sequences: \\n\")\n",
    "        if fileInputPath == '':\n",
    "            \"\"\"This should make it easier to run in Anaconda so that we don't have to keep \n",
    "               restarting the kernel. If enter is pressed but no input is given, we end\n",
    "               the progam.\n",
    "            \"\"\"      \n",
    "            print(\"No input provided. Leaving program....\")\n",
    "            break\n",
    "    # Uses the written getFileExtension function to get the extension from the end of the file\n",
    "        fileExtension = getFileExtension(fileInputPath)\n",
    "    #raises NotFastaOrFastqException if file doesn't end in fasta or fastq\n",
    "        if fileExtension != 'fasta' and fileExtension != 'fastq':\n",
    "            raise NotFastaOrFastqFileError(fileExtension)\n",
    "        try:\n",
    "            \"\"\"Parsing in SeqIO does not appear to return a FileNotFoundError if the file does \n",
    "                not exist. Instead the exception is thrown later. I want to avoid using a lot of try\n",
    "                statements within loops; it seems difficult to keep track of, so below, I am \n",
    "                testing the file beforehand. So a file handle is created beforehand and then closed \"\"\"\n",
    "            fileTest = open(fileInputPath)\n",
    "            fileTest.close()\n",
    "            # if the above statement is successful, we have what we need. We can leave the loop!\n",
    "            break\n",
    "            # if the file is not found, lets us know and allows us the opportunity to try again/.\n",
    "        except FileNotFoundError as e:\n",
    "            print(\"Try again. File not found! \",e)\n",
    "\n",
    "    ### If the file extension is not right, prints out a message, and allows us to try again. \n",
    "    except NotFastaOrFastqFileError as e:\n",
    "        print(\"Try again. This is not a fasta or fastq file, but rather a file of type: \", e)\n",
    "#If we don't specify an input path, we shouldn't bother trying to specify output path. \n",
    "if (fileInputPath != ''):\n",
    "# This allows us to specify wher are outputting the file to. \n",
    "    while True:\n",
    "        try:\n",
    "            # This allows us to specify where we are outputting the file to. \n",
    "            fileOutputPath = input('Write the path of the file you want to output the information to!'\n",
    "                                   \" If this was a mistake, presss enter. Otherwise, provide the file path: \\n'\")\n",
    "\n",
    "            if fileOutputPath == '':\n",
    "                \"\"\"This should make it easier to run in Anaconda so that we don't have to keep \n",
    "                   restarting the kernel. If enter is pressed but no input is given, we end the program. \n",
    "                \"\"\"      \n",
    "                print(\"No output path provided. Leaving program....\")\n",
    "                break\n",
    "            else:\n",
    "                output_file = open(fileOutputPath,'a') \n",
    "                break\n",
    "        except FileNotFoundError:\n",
    "            print(fileOutputPath + \"is not a valid file path. Try again\")\n",
    "### We should stil define fileOutputPath even if we don't have an input path to prevent errors.\n",
    "else:\n",
    "    fileOutputPath = ''\n",
    "\"\"\"we only need to check the fileOutputPath since it is set to empty if fileInputPath is empty.\n",
    "If the fileInputPath is entered, we want to know if the outputPath has been entered. If it has, it is not empty and \n",
    "we can proceed. If it hasn't, it will be empty and we should reach the end of the program. If the fileInputPath is \n",
    "not entered fileOutputPath is set to empty.\"\"\"\n",
    "if (fileOutputPath != ''):\n",
    "    print(\"Outputting information to {0}...\".format(fileOutputPath))\n",
    "            \n",
    "    \"\"\" This creates an iterator that will return SeqRecord objects from the file we have selected, via the path we input from\n",
    "    the interpreter..\n",
    "    This will place less demands on the memory than turning the results of the parse into a list. \n",
    "    fileInputPath and fileExtension are passed from the while loop above after having been\n",
    "    determined to be valid input.\n",
    "    \"\"\"\n",
    "    record_iterator = SeqIO.parse(fileInputPath, fileExtension, IUPAC.ambiguous_dna)\n",
    "    processFile(record_iterator, output_file)\n",
    "    ### This closes the output file    \n",
    "    output_file.close()\n",
    "    print(\"Done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
